File: elixir/day-01/main.exs
defmodule Day1 do
  defp get_digit(line, digits) when byte_size(line) === 0 do
    digits 
  end
  defp get_digit(line, digits) when tuple_size(digits) > 1 do
    head = String.first(line)
    case Integer.parse(head) do
      {_a, _} -> 
        digits = put_elem(digits, 1, head)
        get_digit(String.slice(line, 1..-1//1), digits)
      _ -> get_digit(String.slice(line, 1..-1//1), digits)
    end
  end 

  defp get_digit(line, digits) when tuple_size(digits) < 2 do
    head = String.first(line)
    case Integer.parse(head) do
      {_a, _} -> 
        digits = Tuple.append(digits, head)
        get_digit(String.slice(line, 1..-1//1), digits)
      _ -> get_digit(String.slice(line, 1..-1//1), digits)
    end
  end

  defp combine_tuple(pair) when tuple_size(pair) === 1 do
    combined = elem(pair, 0) <> elem(pair, 0)
    Integer.parse(combined)
  end

  defp combine_tuple(pair) when tuple_size(pair) === 2 do
    combined = elem(pair, 0) <> elem(pair, 1)
    Integer.parse(combined)
  end

  defp get_indices_of_numbers(line, indices, digits, total_digits) when length(digits) === 1 do
    head = hd(digits)
    v = total_digits[head]

    {_, rhead} = Regex.compile(head)
    words = Regex.scan(rhead, line, return: :index)
    words = Enum.map(words, fn t -> {elem(List.first(t), 0), v} end)
    indices = indices ++ words

    {_, rv} = Regex.compile(v)
    nums = Regex.scan(rv, line, return: :index)
    nums = Enum.map(nums, fn t -> {elem(List.first(t), 0), v} end)
    indices = indices ++ nums
    
    Enum.sort_by(indices, fn {value, _} -> value end)
  end

  defp get_indices_of_numbers(line, indices, digits, total_digits) do
    [head | tail] = digits
    v = total_digits[head]

    {_, rhead} = Regex.compile(head)
    words = Regex.scan(rhead, line, return: :index)
    words = Enum.map(words, fn t -> {elem(List.first(t), 0), v} end)
    indices = indices ++ words

    {_, rv} = Regex.compile(v)
    nums = Regex.scan(rv, line, return: :index)
    nums = Enum.map(nums, fn t -> {elem(List.first(t), 0), v} end)
    indices = indices ++ nums

    get_indices_of_numbers(line, indices, tail, total_digits)
  end

  def part1(input, sum) when length(input) > 1 do
    [head | tail ] = input
    {a, _} = get_digit(head, {}) |> combine_tuple()
    sum = sum + a
    part1(tail, sum)
  end

  def part1(input, sum) do
    head = hd(input)
    {a, _} = get_digit(head, {}) |> combine_tuple()
    sum = sum + a
    IO.puts(sum)
  end

  def part2(input, sum, total_digits) when length(input) > 1 do
    [head | tail] = input
    keys = Map.keys(total_digits)
    idxs = get_indices_of_numbers(head, [], keys, total_digits)
    first = List.first(idxs)
    last = List.last(idxs)
    tuple = {elem(first, 1), elem(last, 1)}

    {a, _} = combine_tuple(tuple)
    sum = sum + a
    part2(tail, sum, total_digits)
  end

  def part2(input, sum, total_digits) do
    head = hd(input)
    keys = Map.keys(total_digits)

    idxs = get_indices_of_numbers(head, [], keys, total_digits)
    first = List.first(idxs)
    last = List.last(idxs)
    tuple = {elem(first, 1), elem(last, 1)}

    {a, _} = combine_tuple(tuple)
    sum = sum + a
    IO.puts(sum)
  end
end

  digit_chars = %{
    "one" => "1",
    "two" => "2",
    "three" => "3",
    "four" => "4",
    "five" => "5",
    "six" => "6",
    "seven" => "7",
    "eight" => "8",
    "nine" => "9",
  }

file = File.read!("./input.txt")
file = String.trim(file)
file = String.split(file, "\n")

Day1.part1(file, 0)
Day1.part2(file, 0, digit_chars)

File: rust/day-01/src/main.rs
use std::usize::MAX;

fn main() {
    let input = include_str!("../input.txt");
    let input = input.split("\n").collect::<Vec<&str>>();
    part1(&input);
    part2(&input);
    // println!("{}", input[0]);
}

fn first_and_last_to_number(line: &Vec<char>) -> u32 {
    if line.len() == 0 {
        return 0;
    }

    let mut amount = "".to_string();
    amount.push(*line.first().unwrap());
    amount.push(*line.last().unwrap());
    amount.parse::<u32>().unwrap()
}

fn replace_words_with_digits(line: &str, converter: &Vec<Digit>) -> u32 {
    if line.len() == 0 {
        return 0;
    }

    let contains_word: Vec<(char, Vec<_>)> = converter.iter()
        .filter(|c| line.contains(c.word))
        .map(|c| (c.digit, line.match_indices(c.word).map(|m| m.0).collect())).collect();

    let contains_digit: Vec<(char, Vec<_>)> = converter.iter()
        .filter(|c| line.contains(c.digit))
        .map(|c| (c.digit, line.match_indices(c.digit).map(|m| m.0).collect())).collect();

    // println!("{}", line);
    // println!("{:?}", contains_word);
    // println!("{:?}", contains_digit);

    let mut min = (MAX, '0');
    let mut max = (0, '0');
    for digit in contains_word {
        for idx in digit.1 {
            if idx >= max.0 {
               max = (idx, digit.0); 
            }
            if idx < min.0 {
               min = (idx, digit.0); 
            }
        }
    }

    for digit in contains_digit {
        for idx in digit.1 {
            if idx >= max.0 {
               max = (idx, digit.0); 
            }
            if idx < min.0 {
               min = (idx, digit.0); 
            }
        }
    }
    
    // println!("{:?}", min);
    // println!("{:?}", max);
    first_and_last_to_number(&vec![min.1, max.1])
}

fn part1(input: &Vec<&str>) {
    let digits: Vec<Vec<char>> = input.iter()
    .map(|l| l.chars().filter(|c| c.is_digit(10)).collect())
    .collect();

    let sum: u32 = digits.iter()
        .map(|d| first_and_last_to_number(&d))
        .sum();

    println!("{}", sum);
}

fn part2(input: &Vec<&str>) {
    let converter = vec![
        Digit::new("one", '1'),
        Digit::new("two", '2'),
        Digit::new("three", '3'),
        Digit::new("four", '4'),
        Digit::new("five", '5'),
        Digit::new("six", '6'),
        Digit::new("seven", '7'),
        Digit::new("eight", '8'),
        Digit::new("nine", '9')];

    let sum: u32 = input.iter()
        .map(|d| replace_words_with_digits(&d, &converter))
        .sum();

    println!("{}", sum);
}

struct Digit<'a> {
    word: &'a str,
    digit: char,
}

impl Digit<'_> {
    fn new<'a>(word: &'a str, digit: char) -> Digit<'a> {
        Digit { word, digit }
    }
}


#[test]
fn overlapping() {
    let input = "eightwothree";
    let converter = vec![
        Digit::new("one", '1'),
        Digit::new("two", '2'),
        Digit::new("three", '3'),
        Digit::new("four", '4'),
        Digit::new("five", '5'),
        Digit::new("six", '6'),
        Digit::new("seven", '7'),
        Digit::new("eight", '8'),
        Digit::new("nine", '9')];
    let result = replace_words_with_digits(input, &converter);
    assert_eq!(result, 83);
}

#[test]
fn words_and_numbers() {
    let input = "two1nine";
    let converter = vec![
        Digit::new("one", '1'),
        Digit::new("two", '2'),
        Digit::new("three", '3'),
        Digit::new("four", '4'),
        Digit::new("five", '5'),
        Digit::new("six", '6'),
        Digit::new("seven", '7'),
        Digit::new("eight", '8'),
        Digit::new("nine", '9')];
    let result = replace_words_with_digits(input, &converter);
    assert_eq!(result, 29);
}


#[test]
fn words_and_numbers2() {
    let input = "abcone2threexyz";
    let converter = vec![
        Digit::new("one", '1'),
        Digit::new("two", '2'),
        Digit::new("three", '3'),
        Digit::new("four", '4'),
        Digit::new("five", '5'),
        Digit::new("six", '6'),
        Digit::new("seven", '7'),
        Digit::new("eight", '8'),
        Digit::new("nine", '9')];
    let result = replace_words_with_digits(input, &converter);
    assert_eq!(result, 13);
}

#[test]
fn words_and_numbers3() {
    let input = "xtwone3four";
    let converter = vec![
        Digit::new("one", '1'),
        Digit::new("two", '2'),
        Digit::new("three", '3'),
        Digit::new("four", '4'),
        Digit::new("five", '5'),
        Digit::new("six", '6'),
        Digit::new("seven", '7'),
        Digit::new("eight", '8'),
        Digit::new("nine", '9')];
    let result = replace_words_with_digits(input, &converter);
    assert_eq!(result, 24);
}

#[test]
fn words_and_numbers4() {
    let input = "4nineeightseven2";
    let converter = vec![
        Digit::new("one", '1'),
        Digit::new("two", '2'),
        Digit::new("three", '3'),
        Digit::new("four", '4'),
        Digit::new("five", '5'),
        Digit::new("six", '6'),
        Digit::new("seven", '7'),
        Digit::new("eight", '8'),
        Digit::new("nine", '9')];
    let result = replace_words_with_digits(input, &converter);
    assert_eq!(result, 42);
}

#[test]
fn words_and_numbers5() {
    let input = "zoneight234";
    let converter = vec![
        Digit::new("one", '1'),
        Digit::new("two", '2'),
        Digit::new("three", '3'),
        Digit::new("four", '4'),
        Digit::new("five", '5'),
        Digit::new("six", '6'),
        Digit::new("seven", '7'),
        Digit::new("eight", '8'),
        Digit::new("nine", '9')];
    let result = replace_words_with_digits(input, &converter);
    assert_eq!(result, 14);
}

#[test]
fn with_teens() {
    let input = "7pqrstsixteen";
    let converter = vec![
        Digit::new("one", '1'),
        Digit::new("two", '2'),
        Digit::new("three", '3'),
        Digit::new("four", '4'),
        Digit::new("five", '5'),
        Digit::new("six", '6'),
        Digit::new("seven", '7'),
        Digit::new("eight", '8'),
        Digit::new("nine", '9')];
    let result = replace_words_with_digits(input, &converter);
    assert_eq!(result, 76);
}

#[test]
fn single_digit() {
    let input = "treb7uchet";
    let converter = vec![
        Digit::new("one", '1'),
        Digit::new("two", '2'),
        Digit::new("three", '3'),
        Digit::new("four", '4'),
        Digit::new("five", '5'),
        Digit::new("six", '6'),
        Digit::new("seven", '7'),
        Digit::new("eight", '8'),
        Digit::new("nine", '9')];
    let result = replace_words_with_digits(input, &converter);
    assert_eq!(result, 77);
}


#[test]
fn single_word() {
    let input = "trebsevenuchet";
    let converter = vec![
        Digit::new("one", '1'),
        Digit::new("two", '2'),
        Digit::new("three", '3'),
        Digit::new("four", '4'),
        Digit::new("five", '5'),
        Digit::new("six", '6'),
        Digit::new("seven", '7'),
        Digit::new("eight", '8'),
        Digit::new("nine", '9')];
    let result = replace_words_with_digits(input, &converter);
    assert_eq!(result, 77);
}

#[test]
fn duplicate_digits_and_words() {
    let input = "zoneight21384";
    let converter = vec![
        Digit::new("one", '1'),
        Digit::new("two", '2'),
        Digit::new("three", '3'),
        Digit::new("four", '4'),
        Digit::new("five", '5'),
        Digit::new("six", '6'),
        Digit::new("seven", '7'),
        Digit::new("eight", '8'),
        Digit::new("nine", '9')];
    let result = replace_words_with_digits(input, &converter);
    assert_eq!(result, 14);
}

